### PR TITLE
Improvements to the step logs storage functionality

### PR DESCRIPTION
**Pull request by bcdurak**
## Describe changes

This PR introduces a patch to the step-logging storage functionality. The most significant changes include:

- The storage of the logs and outputs is now completely disconnected from the `logging` module.
- This means, once step log storage is enabled, the `stdout` and `stderr` outputs are no longer converted to logging messages.
- Through the second point, outputs (like progress bars) from libraries such as `tdqm` or `keras` are now displayed properly.
- In order to save any output (such as `print` statements, any `logging` messages, etc. ), we now patch the `stdout` and `stderr` and wrap their `write` methods. This wrapper is responsible for storing the outputs as step logs.
- There is an additional environmental variable called `ZENML_DISABLE_STEP_LOGS_STORAGE` which takes precedence over the other ways of enabling/disabling this feature.
- Previously, this feature was only enabled in the `StepLauncher` interface. With the new changes, it is now also enabled on the `StepRunner` interface. 
- This means, if your step is running in a different environment than the orchestration environment, the logs will still get saved in the artifact store.

## Code for manual test

```python
import logging
import sys
from time import sleep
from tqdm import tqdm, trange

from zenml import pipeline, step


@step(enable_cache=False)
def logging_step() -> None:
    logging.info('Info message!')
    logging.warning('Warning message!')
    logging.error('Error message!')


@step(enable_cache=False)
def tqdm_step() -> None:
    text = ""
    for char in tqdm(["a", "b", "c", "d"]):
        sleep(0.25)
        text = text + char

    for _ in trange(100):
        sleep(0.01)

    with tqdm(total=100) as pbar:
        for i in range(10):
            sleep(0.1)
            pbar.update(10)


@step(enable_cache=False)
def print_step() -> None:
    print("Print test")
    print("Print test with stderr", file=sys.stderr)


@pipeline()
def my_pipeline():
    logging_step()
    tqdm_step()
    print_step()


if __name__ == "__main__":
    my_pipeline()
```

## Remaining TODOs:

- Update the docs with the new environmental variable and screenshots.
- Linting, formatting, docstrings...

## Pre-requisites
Please ensure you have done the following:
- [X] I have read the **CONTRIBUTING.md** document.
- [ ] If my change requires a change to docs, I have updated the documentation accordingly.
- [X] If I have added an integration, I have updated the [integrations](https://docs.zenml.io/component-gallery/integrations) table and the [corresponding website section](https://zenml.io/integrations).
- [ ] I have added tests to cover my changes.

## Types of changes

- [X] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Other (add details above)

### CODE REVIEW

**Review by stefannica — CHANGES_REQUESTED** [review]
I just love how this simplifies the log duplication ! I left some questions and suggestions for improvement, but nothing major otherwise.

File: src/zenml/orchestrators/step_runner.py [hunk 1]
```diff
@@ -105,108 +111,141 @@ def run(
...
             )
 
-            step_failed = False
-            try:
-                return_values = step_instance.call_entrypoint(
-                    **function_params
+        logs_context = nullcontext()
+        if step_logging_enabled and not redirected.get():
+            if step_run.logs:
+                logs_context = StepLogsStorageContext(  # type: ignore[assignment]
```
[the remark this answers is not in the document]
Comment (bcdurak) [reply src/zenml/orchestrators/step_runner.py:133]: Step launchers run in the orchestration environment, whereas step runners run in the step environment. 

For orchestrators, where these two environments are not the same, the code in the step launcher takes care of information such as `Step X has started/ended in X seconds.`, whereas the step runner takes care of the logs and prints happening during the step execution.

The `redirected` variable takes care of the instances (such as the local orchestration) where the orchestration environment is the same as the step execution environment and makes sure that the logs are not duplicated.

Hope that clears it.

File: src/zenml/logging/step_logging.py [hunk 2]
```diff
@@ -76,138 +87,64 @@ def prepare_logs_uri(
...
                 self.disabled = True
-                with fileio.open(self.logs_uri, mode="wb") as log_file:
-                    log_file.write(self.buffer.getvalue().encode("utf-8"))
+
+                if self.buffer:
+                    with fileio.open(self.logs_uri, "a") as file:
+                        for message in self.buffer:
+                            file.write(
+                                remove_ansi_escape_codes(message) + "\n"
+                            )
```
[the remark this answers is not in the document]
Comment (bcdurak) [reply src/zenml/logging/step_logging.py:144]: You are correct. Currently, the `logger` instances created through ZenML's `get_logger` function add all sorts of ANSI escape codes to the raw messages. Adding support to the Dashboard is a good idea, but until that is implemented, I would leave this as is.

File: src/zenml/logging/step_logging.py [hunk 3]
```diff
@@ -76,138 +87,64 @@ def prepare_logs_uri(
...
+            text: the incoming string.
         """
-        msg = self.format(record)
-        self.buffer.write(msg + "\n")
-        self.message_count += 1
-
-        current_time = time.time()
-        time_elapsed = current_time - self.last_upload_time
+        if text == "\n":
+            return
```
[the remark this answers is not in the document]
Comment (bcdurak) [reply src/zenml/logging/step_logging.py:124]: At the moment, I am not sure where the root is, but if I do not apply this condition, the `Storage` instance is getting more `"\n"` messages than initially expected. I have seen the same implementation in a few more examples of the same use-case, thus applied it here as well.

File: src/zenml/logging/step_logging.py [hunk 4]
```diff
@@ -217,47 +154,65 @@ def flush(self) -> None:
...
-    Args:
-        logs_uri: The URI of the output artifact.
+        Returns:
+            self
+        """
+        self.stdout_write = getattr(sys.stdout, "write")
+        self.stderr_write = getattr(sys.stderr, "write")
+
+        setattr(sys.stdout, "write", self._wrap_write(self.stdout_write))
+        setattr(sys.stderr, "write", self._wrap_write(self.stdout_write))
```
[the remark this answers is not in the document]
Comment (bcdurak) [reply src/zenml/logging/step_logging.py:182]: You are right. I will look into this.

[the remark this answers is not in the document]
Comment (bcdurak) [reply src/zenml/logging/step_logging.py]: Implemented and pushed.

File: src/zenml/logging/step_logging.py [hunk 5]
```diff
@@ -76,188 +86,167 @@ def prepare_logs_uri(
...
-        if message != "\n":
-            self.stdout_logger.info(message)
-        return len(message)
+        if text == "\n":
+            return
 
+        if not self.disabled:
+            self.buffer.append(text)
+        else:
+            self.disabled_buffer.append(text)
```
Comment (stefannica) [inline src/zenml/logging/step_logging.py:129]: I don't think you should use a separate buffer just to store messages that are generated by the process that saves logs to the artifact store. Those logs should basically never be stored in the artifact store. In other words, there should be no `else` here.

**Review by stefannica — CHANGES_REQUESTED** [review]

File: src/zenml/logging/step_logging.py (hunk 5 above)
Comment (bcdurak) [reply src/zenml/logging/step_logging.py]: removed the separate buffer.

**Review by stefannica — APPROVED** [review]

**Pull request merged**